File: Leetcode 17 - letter-combinations-of-a-phone-number/17.letter-combinations-of-a-phone-number.cpp
/*
 * @lc app=leetcode id=17 lang=cpp
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  vector<string> letterCombinations(string digits) {
    if (digits.empty())
      return {};

    vector<string> result;
    string path;
    backtrack(digits, 0, result, path);
    return result;
  }

private:
  void backtrack(string &digits, int start, vector<string> &result,
                 string &path) {
    if (start == digits.size()) {
      result.push_back(path);
      return;
    }

    string letters = getLetters(digits[start]);
    for (char letter : letters) {
      path.push_back(letter);
      backtrack(digits, start + 1, result, path);
      path.pop_back();
    }
  }

  string getLetters(char digit) {
    switch (digit) {
    case '2':
      return "abc";
    case '3':
      return "def";
    case '4':
      return "ghi";
    case '5':
      return "jkl";
    case '6':
      return "mno";
    case '7':
      return "pqrs";
    case '8':
      return "tuv";
    case '9':
      return "wxyz";
    default:
      return "";
    }
    return "";
  }
};

int main() {
  Solution sol;

  string digits = "23";
  vector<string> result = sol.letterCombinations(digits);
  for (const auto &comb : result) {
    cout << comb << endl; // ad ae af bd be bf cd ce cf
  }

  digits = "2";
  result = sol.letterCombinations(digits);
  for (const auto &comb : result) {
    cout << comb << endl; // a b c
  }
}
// @lc code=end

File: Leetcode 17 - letter-combinations-of-a-phone-number/17.letter-combinations-of-a-phone-number.java
/*
 * @lc app=leetcode id=17 lang=java
 *
 * [17] Letter Combinations of a Phone Number
 */

// @lc code=start

import java.util.ArrayList;
import java.util.List;

class Solution {

  private List<String> result = new ArrayList<>();
  private String[] digitToChar = {
    "",
    "",
    "abc",
    "def",
    "ghi",
    "jkl",
    "mno",
    "pqrs",
    "tuv",
    "wxyz",
  };

  public List<String> letterCombinations(String digits) {
    if (digits.isEmpty()) return result;

    backtrack(digits, new StringBuilder(), 0);
    return result;
  }

  private void backtrack(String digits, StringBuilder curStr, int cur) {
    if (curStr.length() == digits.length()) {
      result.add(curStr.toString());
      return;
    }

    String chars = digitToChar[digits.charAt(cur) - '0'];
    for (char c : chars.toCharArray()) {
      curStr.append(c);
      backtrack(digits, curStr, cur + 1);
      curStr.deleteCharAt(curStr.length() - 1);
    }
  }

  public static void main(String[] args) {
    Solution solution = new Solution();

    for (String s : solution.letterCombinations("23")) {
      System.out.println(s);
    } // ["ad","ae","af","bd","be","bf","cd","ce","cf"]
    for (String s : solution.letterCombinations("")) {
      System.out.println(s);
    } // []
  }
}
// @lc code=end
